### PR TITLE
Gutenboarding: Send "guessed" browser timezone string for new sites

### PR DESCRIPTION
**Pull request by p-jackson**
#### Changes proposed in this Pull Request

The site creation used in production now creates sites with a best guess of what their timezone should be (#38891). This does the same thing, but in gutenboarding.

* Send `timezone_string` param when creating new site in gutenboarding
* Guess timezone using code from `lib/i18n-utils/guess-timezone`
* Remove `moment-timezone` dependency from `guessTimezone()` so we don't have to include it in gutenboarding
* Update existing uses of `guessTimezone()` so they deal with possible failure since `Intl` API may not be available (it is in all the browsers we support, but we still want the page to mostly work on unsupported browsers). If `Intl` is unavailable:
  * when creating a site, then the `timezone_string` param isn't sent
  * when viewing a site's general settings page then we don't suggest a timezone
  * when booking a concierge session then the form defaults to UTC

This PR pulls in non-packaged code from Calypso which isn't ideal, given we're trying to keep Calypso dependencies out of gutenboarding. However `i18n-utils` is imported in a few places already, so I think it'd be good to come up with a generic solution that fixes all of these imports at once.

#### Testing instructions

* Create a new site via `/gutenboarding`
* Check the API request to `/sites/new` contains the correct `timezone_string` param
* The site's time zone should be that of the computer used to create the site (visible at `/settings/general/{site_slug}`)

Testing other uses of `guessTimezone()` haven't broken:

* Create a new site starting from `/start` and check the site's timezone is correct
* The timezone suggestion button at `/settings/general/{siteslug}` should still work
* The timezone option at the conceirge signup form should still default to the guessed timezone `/me/concierge/{siteslug}/book`

Can't think of a way to test what happens when `Intl` isn't available in a browser other than hardcode `guessTimezone()` to always return `undefined`.

### CODE REVIEW

**Comment by p-jackson** [comment]
> I know there've been some efforts to remove `moment` where it isn't needed to improve load times

Interesting. I know `moment` can be pretty heavy. Thing is we don't need this to do the initial render, so it might be something to delay laod.

**Comment by andrewserong** [comment]
I'm not totally across it, but this is one of the recent PRs I saw on the topic: https://github.com/Automattic/wp-calypso/pull/39483

**Comment by simison** [comment]
Yes, @sgomes has been working hard to remove it from everywhere. :-)

**Comment by p-jackson** [comment]
Internally `moment` is using `Intl.DateTimeFormat` if it's available, and falling back to trickier code if it's not there.

Looks like the browsers we're interested in support `Intl.DateTimeFormat`: `https://caniuse.com/#search=DateTimeFormat

So I think I'll switch to using `DateTimeFormat` directly if it's available, and return nothing if it's not.

**Comment by ramonjd** [comment]
> So I think I'll switch to using DateTimeFormat directly if it's available, and return nothing if it's not.

This is a good plan. Thanks for picking this one up.

```
new Intl.DateTimeFormat().resolvedOptions().timeZone
> "Australia/Sydney"
```

👍 

Wonder if we'll have to do any [rewriting of timezone output as we do now](https://github.com/Automattic/wp-calypso/blob/46da258/client/lib/i18n-utils/guess-timezone.js#L17) (?)

**Comment by sgomes** [comment]
Thank you for helping out in reducing our dependency on `moment`! 🙂

**Comment by p-jackson** [comment]
> Wonder if we'll have to do any rewriting of timezone output as we do now (?)

Yeah I think we're going to need to keep that rewriting code. Would be nice to share it somehow.

**Comment by ramonjd** [comment]
> Yeah I think we're going to need to keep that rewriting code. Would be nice to share it somehow.

One day create a `i18n-utils` package? Or under or `packages/i18n-calypso/utils` ?

Works great for me!!

### Before

Screen Shot 2020-03-30 at 10 45 13 am


### After

Screen Shot 2020-03-30 at 10 55 54 am
Screen Shot 2020-03-30 at 1 57 05 am

**Comment by p-jackson** [comment]
I've removed the `moment-timezone` dependency. I decided to remove the dependency from the Calypso code too, which does make this a bigger change, but figured I'd refactoring to share the `linkedTimezones` data anyway.
Luckily timezone guessing was only being used for optional things, so it's easy to fail gracefully if the browesr doesn't support `Intl`.

It looks like it's made a big impact if you check at [ICFY](http://iscalypsofastyet.com/branch?branch=add/timezone-guess-creating-site), if I'm interpretting correctly. @sgomes is it possible to make ICFY report results in the PR thread again? Just so it's easier for me to compare, I'm more used to the way it talks about app entries.

File: client/lib/i18n-utils/guess-timezone.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	return timezone;
+}
+
+function isLinkedTimezone( timezone: string ): timezone is keyof typeof linkedTimezones {
+	return timezone in linkedTimezones;
+}
+
+export default function guessTimezone() {
```
Comment (ramonjd) [inline client/lib/i18n-utils/guess-timezone.ts]: Typescript n00b question: is there a way to handle differing return types? Like `: string | undefined` or something?

Comment (p-jackson) [reply client/lib/i18n-utils/guess-timezone.ts]: Yeah that's how you'd set the return type explicitly. If you hover over `guessTimezone()` you'll see that's what TypeScript has inferred.

I'm usually in two minds about whether to explicitly set the return type or not. In this case it'd probably be good.

Comment (ramonjd) [reply client/lib/i18n-utils/guess-timezone.ts]: **TIL**

🙇

File: client/lib/i18n-utils/guess-timezone.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
+// This list comes from https://github.com/eggert/tz/blob/master/backward
+const linkedTimezones = {
```
Comment (alshakero) [inline client/lib/i18n-utils/guess-timezone.ts]: Would it make sense to export this data from [this file](https://github.com/Automattic/wp-calypso/blob/master/client/lib/i18n-utils/guess-timezone.js#L17) and import it here instead? This add a redundant ~1KB to the uncompressed bundle.

**Comment by sgomes** [comment]
> @sgomes is it possible to make ICFY report results in the PR thread again? Just so it's easier for me to compare, I'm more used to the way it talks about app entries.

That should still be working (and it is, from what I see). You only have to scroll up and find [the original entry](https://github.com/Automattic/wp-calypso/pull/40508#issuecomment-604802425), which should get updated when new commits are added. Sometimes there can be a bit of lag, depending on how busy the build queue is.

**Comment by sgomes** [comment]
Also, the current numbers are looking quite good! 👍

**Comment by p-jackson** [comment]
I see the update now. Sweet! Happy with that :)

**Pull request merged**